src/bin/sccache-dist/main.rs

```diff
@@ -668,9 +675,77 @@ impl SchedulerIncoming for Scheduler {
         let mut servers = self.servers.lock().unwrap();
 
         match servers.get_mut(&server_id) {
-            Some(ref mut details) if details.server_nonce == server_nonce => {
-                details.last_seen = Instant::now();
-                return Ok(HeartbeatServerResult { is_new: false });
+            Some(ref mut details) => {
+                if details.server_nonce == server_nonce {
+                    let now = Instant::now();
+                    details.last_seen = now;
+
+                    let mut stale_jobs = Vec::new();
+                    for (&job_id, &last_seen) in details.jobs_unclaimed.iter() {
+                        if now.duration_since(last_seen) > UNCLAIMED_READY_TIMEOUT {
+                            let jobs = self.jobs.lock().unwrap();
+                            if let Some(detail) = jobs.get(&job_id) {
+                                match detail.state {
+                                    JobState::Ready => {
+                                        stale_jobs.push(job_id);
+                                    },
+                                    JobState::Pending => {
+                                        if now.duration_since(last_seen) >
+                                            UNCLAIMED_PENDING_TIMEOUT {
+                                                stale_jobs.push(job_id);
+                                            }
+                                    },
+                                    state => {
+                                        warn!("Invalid unclaimed job state for {}: {}",
+                                              job_id, state);
+                                    }
+                                }
+                            } else {
+                                warn!("Unknown stale job {}", job_id);
+                                stale_jobs.push(job_id);
+                            }
+                        }
+                    }
+
+                    if stale_jobs.len() > 0 {
+                        warn!("The following stale jobs will be de-allocated: {:?}",
+                              stale_jobs);
+
+                        let mut jobs = self.jobs.lock().unwrap();
+
+                        for job_id in stale_jobs {
+                            if !details.jobs_assigned.remove(&job_id) {
```

Comment:
Is it possible to define a local enum or something so we can push (job_id, enumstate) triples into `stale_jobs`, and then we know exactly where to remove the jobs from in this loop?  I guess we have to worry about jobs appearing in `stale_jobs` from above and then getting removed before we get here?

Reply:
I don't think I quite understand this suggestion. We want to remove the job from all these places whenever we find one that's expired. All this conditional checking is just to log something in case we think there should be a job somewhere and it isn't. What would the enumstate tell us?

Reply:
OK, I see now I was misunderstanding what was going on here.